pkg/gatewayserver/gatewayserver.go

```diff
@@ -604,7 +604,16 @@ func (gs *GatewayServer) startDisconnectOnChangeTask(conn connectionEntry) {
 				},
 			})
 			if err != nil {
-				log.FromContext(ctx).WithError(err).Warn("Failed to get gateway")
+				if errors.IsUnauthenticated(err) || errors.IsPermissionDenied(err) {
+					// Since there is an active connection, the `Get` request will not return a `NotFound` error as the gateway existed during the connect, since the rights assertion fails first.
+					// Instead,
+					// 1. If the gateway is connected with an API key and is deleted, the IS returns an `Unauthenticated`, since the API Key is also deleted.
+					// 2. If the gateway is connected without an API key (UDP, LBS in unauthenticated mode) and is deleted the IS returns an `PermissionDenied` as there are no rights for these IDs.
```

Comment:
I don't get this. If there is no authentication to forward, GS uses cluster authentication (https://github.com/TheThingsIndustries/lorawan-stack/blob/v3.16/pkg/gatewayserver/is.go#L67-L69). IS will just return `NotFound` right?

Reply:
This is quite non-intuitive. I had the same expectation as you did and was surprised when the tests failed. So here it goes
1. On `Get`, the IS asserts the rights using [`RequireGateway`](https://github.com/TheThingsNetwork/lorawan-stack/blob/a053461febe1813ad537ef7ba38533e026c7b6cc/pkg/auth/rights/require.go#L159)
2. The IS then looks for the gateway store with the IDs. It returns a [`NotFound`](https://github.com/TheThingsNetwork/lorawan-stack/blob/c2ce76a770e8ac689a63508711e51e16a1519af2/pkg/identityserver/rights.go#L166-L170). Since we are using cluster auth and not doing a `Get` as admin, the cause of the error (NotFound) is suppressed and this function returns empty rights.
3. Now back in `RequireGateway`, since the returned rights are empty, we return a [PermissionDenied](https://github.com/TheThingsNetwork/lorawan-stack/blob/a053461febe1813ad537ef7ba38533e026c7b6cc/pkg/auth/rights/require.go#L165-L167).

I think the reason for this behaviour is to not leak whether a gateway is present or not for someone who wants to scan the database.